### PR TITLE
Integrate pallet proxy

### PR DESCRIPTION
**Pull request by Moliholy**

### CODE REVIEW

**Comment by peterwht** [comment]
It is looking good! 

As discussed, add the following `ProxyType`
```
pub enum ProxyType {
	/// All calls can be proxied. This is the trivial/most permissive filter.
	Any,
	/// Only extrinsics that do not transfer funds.
	NonTransfer,
	/// Only extrinsics related to governance (democracy and collectives).
	Governance,
	/// Allow to veto an announced proxy call.
	CancelProxy,
	/// Allow extrinsic related to Balances.
	Balances,
	/// Allow extrinsic related to DID pallet
	DidManagement
}
```

- Update the instance filter with the new `ProxyType`.
- Add to `define_benchmarks` and run (locally for now)

**Comment by Moliholy** [comment]
@peterwht I just finished the implementation. Feel free to review it.

File: runtime/devnet/src/lib.rs [hunk 1]
```diff
@@ -477,6 +483,86 @@ parameter_types! {
...
+impl InstanceFilter<RuntimeCall> for ProxyType {
+	fn filter(&self, c: &RuntimeCall) -> bool {
+		match self {
+			ProxyType::Any => true,
+			ProxyType::NonTransfer =>
+				!matches!(c, RuntimeCall::Balances(..) | RuntimeCall::Assets(..)),
+			ProxyType::Governance =>
+				matches!(c, RuntimeCall::Council(..) | RuntimeCall::Treasury(..)),
+			ProxyType::CancelProxy =>
+				matches!(c, RuntimeCall::Proxy(pallet_proxy::Call::remove_announcement { .. })),
```
Comment (peterwht) [inline runtime/devnet/src/lib.rs:536]: Should this be `reject_announcement`?

See statemine config: https://github.com/paritytech/cumulus/blob/polkadot-v0.9.30/parachains/runtimes/assets/statemine/src/lib.rs

And polkadot: https://github.com/paritytech/polkadot/blob/release-v0.9.30/runtime/polkadot/src/lib.rs

File: runtime/devnet/src/lib.rs [hunk 2]
```diff
@@ -477,6 +483,86 @@ parameter_types! {
...
+
+impl pallet_proxy::Config for Runtime {
+	type RuntimeEvent = RuntimeEvent;
+	type RuntimeCall = RuntimeCall;
+	type Currency = Balances;
+	type ProxyType = ProxyType;
+	type ProxyDepositBase = ProxyDepositBase;
+	type ProxyDepositFactor = ProxyDepositFactor;
+	type MaxProxies = ConstU32<32>;
+	type WeightInfo = pallet_proxy::weights::SubstrateWeight<Runtime>;
```
Comment (peterwht) [inline runtime/devnet/src/lib.rs:559]: Run benchmark (locally is fine) and update `WeightInfo` to use generated file.

File: runtime/devnet/src/lib.rs (hunk 1 above)
Comment (Moliholy) [reply runtime/devnet/src/lib.rs:536]: It seems so, but I got confused by the documentation:

**reject_announcement:**
```
Remove the given announcement of a delegate.
May be called by a target (proxied) account to remove a call that one of their delegates (delegate) has announced they want to execute. The deposit is returned.
```

**remove_announcement:**
```
Remove a given announcement.
May be called by a proxy account to remove a call they previously announced and return the deposit.
```

First one says it is supposed to be called by the original account, not by a proxy. The second one states the opposite, and yet we are using the first one. It sounds counter-intuitive, and that's why I got confused.

Anyhow, I just updated it according to your suggestion.

File: runtime/devnet/src/lib.rs (hunk 2 above)
Comment (Moliholy) [reply runtime/devnet/src/lib.rs]: Done!

File: runtime/devnet/src/lib.rs (hunk 1 above)
Comment (peterwht) [reply runtime/devnet/src/lib.rs]: Awesome! I believe it is intended so a proxy account, that has another proxy, can reject calls that the delegate account has proposed.

**Review by peterwht — APPROVED** [review]

**Review by niall-watr — APPROVED** [review]

**Pull request merged**